crates/bazed-core/src/vim_interface.rs

```diff
@@ -16,176 +25,339 @@ pub enum VimMode {
+        ),
+        (
+            key("x").into(),
+            leaf("delete", |v, b, _, _| {
+                b.apply_buffer_op(&v.vp, BufferOp::DeleteSelected);
+            }),
+        ),
+    ]));
+    visual_mode_movement.merge(keymap)
+}
 
-    pub(crate) fn on_input_visual_mode(
-        &mut self,
-        view: &View,
-        buffer: &mut Buffer,
-        input: &KeyInput,
-    ) {
-        if let Some(motion) = key_to_motion_normal(input.ctrl_held(), &input.key) {
-            buffer.apply_buffer_op(&view.vp, BufferOp::Selection(motion));
-            return;
-        }
-        match input.key {
-            Key::Escape => {
-                self.mode = VimMode::Normal;
-                buffer.collapse_selections();
-            },
-            Key::Char('d') | Key::Char('x') => {
-                buffer.apply_buffer_op(&view.vp, BufferOp::DeleteSelected);
-            },
-            _ => {
-                if let Some(motion) = key_to_motion(input.ctrl_held(), &input.key) {
-                    buffer.apply_buffer_op(&view.vp, BufferOp::Selection(motion));
-                }
-            },
-        }
-    }
+fn add_caret_keymap() -> Keymap<MappedFn> {
+    normal_mode_movement_key_motion_keymap().map(&|motion: Motion| {
+        mapping(move |v, b, _, _| b.apply_buffer_op(&v.vp, BufferOp::NewCaret(motion)))
+    })
+}
 
-    pub(crate) fn on_input_replace_mode(
-        &mut self,
-        view: &View,
-        buffer: &mut Buffer,
-        input: &KeyInput,
-    ) {
-        if self.on_movement_key(view, buffer, input) {
-            return;
-        }
-        match input.key {
-            Key::Escape => {
-                self.mode = VimMode::Normal;
-            },
-            Key::Char(c) => buffer.replace_at_carets(&c.to_string()),
-            _ => {},
-        }
-    }
+fn normal_mode_movement_key_keymap() -> Keymap<MappedFn> {
+    normal_mode_movement_key_motion_keymap().map(&|motion: Motion| {
+        mapping(move |v, b, _, _| b.apply_buffer_op(&v.vp, BufferOp::Move(motion)))
+    })
+}
 
-    /// Handle regular movement keys (usable for normal and insert mode)
-    fn on_movement_key(&mut self, view: &View, buffer: &mut Buffer, input: &KeyInput) -> bool {
-        let Some(motion) = key_to_motion(input.ctrl_held(), &input.key)  else {
-            return false;
-        };
-        let op = if input.alt_held() {
-            BufferOp::NewCaret(motion)
-        } else {
-            BufferOp::Move(motion)
-        };
-        buffer.apply_buffer_op(&view.vp, op);
-        true
-    }
+fn movement_key_keymap() -> Keymap<MappedFn> {
+    movement_key_motion_keymap()
+        .map(&|motion| mapping(move |v, b, _, _| b.apply_buffer_op(&v.vp, BufferOp::Move(motion))))
+}
 
-    fn switch_mode(&mut self, mode: VimMode) {
-        self.mode = mode;
-    }
+fn normal_mode_movement_key_motion_keymap() -> Keymap<Motion<'static>> {
+    let normal_mode_movement_binds = Keymap::new_from_map(HashMap::from_iter([
```

Comment:
Eh, maplit annoying because it breaks auto formatting and LSP. My plan would be to try to either have some builder api or something here that's clean enough to read, _or_ to just write a _proper_ proc macro for this... but until then, eventhough this reads kinda badly, the lsp is worth it